pallets/identity/src/lib.rs

```diff
@@ -145,5 +158,20 @@ pub mod pallet {
 
             Ok(())
         }
+
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        /// Issue a signed Fennel signal on behalf of an owned identity.
+        pub fn sign_for_identity(origin: OriginFor<T>, identity_id: u32, content: Vec<u8>) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            ensure!(Self::is_identity_owned_by_sender(&who, &identity_id), Error::<T>::IdentityNotOwned);
+            let signal_id: u32 = <SignalCount<T>>::get();
+            <FennelSignal<T>>::insert(&identity_id, &signal_id, &content);
+            let new_id: u32 = signal_id.checked_add(1).ok_or(Error::<T>::StorageOverflow)?;
+            <SignalCount<T>>::put(new_id);
+            Self::deposit_event(Event::SignedSignal(identity_id, who, content));
+
+            Ok(())
+        }
```

Comment:
are we assuming the content to already be signed? or are we going to be adding the signing functionality later?

Reply:
Transactions are signed by definition once they're submitted to the chain. We have two options here - just rely on the SubKey signature, or generate a signature based on the key used by the application issuing the signal. Just using the SubKey signature would take up less space and therefore less time.

Reply:
Any downside to just relying on that? Any upside to generating a signature based on the key being used by the application?

Reply:
Good question. The downside would be that the signature would *always* come from the account, rather than the identity, which might not always be ideal. That leads us to the upside of using the identity's key, which is that it's directly connected back to that identity and not the SubKey that owns the identity.

If we generate a signature with the _identity's_ key, or the key being used by the application, then our downsides are additional size to the transaction and the slight possibility of apps not managing the keys for different identities correctly. That's not really our problem though so it should be fine. I think it's dependent on context, honestly, sometimes you might want one or the other, which is why I wanted to open this discussion.

Reply:
In my experience, tying transactions to an account, rather than an identity, inevitably leads to an expensive operation and complex problem to solve _later_ when you inevitably need to walk your way to obtaining the identity tied to a transaction. You might save on the size of the transaction now, but will likely create tech debt to solve later. Seems like a matter of picking the trade-off you want to live with. No perfect solution based upon my understanding of the choices.

Reply:
I think you're absolutely right. I'm leaning heavily toward identity.

Reply:
Yeah, philosophically, I generally prefer to pay bills of known size now, rather than bills of unknown size later -- unless there is a really good and compelling reason to not do that.

Reply:
The only reason right now is that we don't have decentralized key management implemented. That's milestone 2.

Reply:
And frankly, we'll need to offload signature processing to applications. You don't want that happening during consensus.

Reply:
I want to step back further and get on the same page, conceptually. I find this helps me think more clearly about what we mean by different words and (more importantly) really drives home what it is we are trying to accomplish in any given scenario.

I apologize for the length of this comment. I have been following the conversation and generated a lot of thoughts in the meantime.

Also, I am mostly writing this out so that you guys can criticize the gaps in my knowledge. You guys already know a lot of what I am saying here -- but I don't know what I don't know -- and so by writing out all my thoughts (the extent of my knowledge related to this topic) you guys can now fill in the gaps which gives me the opportunity to learn.

With that said, what do we mean by "identity"? In a very crude sense, we could mean "the unique u32 index which lives inside an array which is mapped to an AccountID". That would be technically correct, but the Fennel concept of "identity" is more than that.

An identity, as far as I understand, is meant to be a vehicle for storing information that we want to define that identity (ie a name, a date of birth, a job title -- whatever is relevant for that identity).

An identity is also meant to be used as a tool for proof -- ie, yes that is me and here is how I can prove it to you: signs a request for proof with private key (which only I have) and the other can then validate against our public key.

My knowledge is limited at this point. I may not see the full picture here and would like input on where my understanding is falling short.

With all that said, the question I believe we are discussing is: do we need to sign every piece of information with the private key associated with the identity through which it is coming through?

Right now, we do not having anything setup which generates public/private keys for every identity created under an AccountID.

This might be counter-intuitive, but I wonder what the direct benefits of generating on-chain public/private key for every identity?

Conceptually, it makes sense that we would want every identity to be _unique_ -- how else is this done except by generating public/private keys for each identity?

We must ask ourselves, however, what is gained by doing this, what is the cost, and what alternatives exist which accomplish the same goal of tying information to an identity and using it as a tool of proof.

Let's say we want to join a community, and they require a Fennel identity to participate. Presumably, all applications participating in the Fennel network will require a Fennel identity (ie signup for google to use gmail -- give us an identity that you own to use our Fennel app).

That being the case, every Fennel app will send a request that asks for the identity you want to use. You will "prove" that you are the owner of that identity by virtue of the transaction occurring -- the transaction will fail if the AccountID of the sender (you responding to the request from the app) does not own the Identity being given to the request.

To this end, a unique public/private key for the Identity is not required -- only rely on the nature of how transactions operate (as @Romulus10 mentioned earlier) and the runtime logic we add which will ensure the given u32 index which represents an identity is actually owned by the sender's AccountID.

What about adding information to an identity? How do you prove that information is tied to a particular identity?

By virtue of the information being on-chain, mapped to an identity which is mapped to an AccountID -- that is all the proof you need. The information would not have been allowed to exist there unless the transactions that placed the information there were signed by the AccountID which owns the identity.

There might be more to the story that I am missing. But based on all of this, it appears we do not need to generate public/private keys for each identity.

Unless, of course, I am about to be schooled by you guys ;)

Reply:
So let me respond to that by starting at the top, then we can break it down further - What is an identity? On Fennel Protocol, an Identity is a collection of information logically bound to a single point owned by a Fennel account. So an Identity might be a record connecting your account to a public key and email address. It might be a record connecting you to an application account. It's a bit fuzzy of a word, I will definitely admit that, but it captures the idea of a point of reference for the rest of the network to interact with the owner in certain contexts.

Reply:
There's a ton of good commentary in your comment @isaacadams, and I want to make sure we tackle it all in sufficient detail.

Reply:
I should also comment now to say that the keys are STORED by Fennel Protocol, but should in no circumstances be GENERATED by the protocol. We haven't gotten this far yet.

Reply:
At the very least, users should have the OPTION to generate a key for each identity. I do see where Isaac is going with his logic and it had occurred to me as well. It would save space, time, and effort to rely on signatures as presented by the transaction.

Reply:
> At the very least, users should have the OPTION to generate a key for each identity.

would this be handled at the application level?

Reply:
Yes. Anything super expensive we're going to want to handle off-chain.